File: solutions/Leetcode_95/leetcode_95.cpp
#include <vector>

class Solution1
{
public:
    std::vector<TreeNode*> generateTrees(int n)
    {
        return _generateTreesInRange(1, n);
    }

private:
    std::vector<TreeNode*> _generateTreesInRange(int m, int n)
    {
        if (m > n)
           return {nullptr};
        std::vector<TreeNode*> trees;
        for (int x = m; x <= n; x++)
            for (TreeNode* leftTree : _generateTreesInRange(m, x-1))
                for (TreeNode* rightTree : _generateTreesInRange(x+1, n))
                {
                    TreeNode* node = new TreeNode(x);
                    node->left = leftTree;
                    node->right = rightTree;
                    trees.push_back(node);
                }
        return trees;
    }
};

File: solutions/Leetcode_95/leetcode_95.py
from functools import lru_cache
from typing import List, Optional

from leetcode.python.tree import TreeNode


class Solution1:
    def generate_trees(self, n: int) -> List[Optional[TreeNode]]:
        return self._generate_trees_in_range(1, n)

    @lru_cache
    def _generate_trees_in_range(self, m, n):
        if m > n:
            return [None]
        trees = []
        for x in range(m, n+1):
            left_trees = self._generate_trees_in_range(m, x-1)
            right_trees = self._generate_trees_in_range(x+1, n)
            for left_tree in left_trees:
                for right_tree in right_trees:
                    node = TreeNode(x)
                    node.left = left_tree
                    node.right = right_tree
                    trees.append(node)
        return trees
